Implement method to bulk add all collection elements to a PriorityQueue

BaurzhanSakhariev opened the pull request:
Hi!

This PR doesn't have a link to a Jira Issue as according to contribution guidelines I can either create a patch and attach it to an issue on Jira or open a pull request - let me know if I misunderstood things and have to create a Jira issue anyway.


# Description

`PriorityQueue` class doesn't have a constructor accepting array/collection and I created one for the cases, when heap elements are known in advance to save some time on queue creation: `O(N)` ([heapify](https://en.wikipedia.org/wiki/Binary_heap#Building_a_heap)) vs `Nlog(N)` (calling add() in loop).

# Solution

JDK `PriorityQueue` has a constructor, accepting Collection and it internally uses `heapify` - this PR is basically adaptation of this method (index is not 0 but 1 based and no need to check comparator presence).

# Tests

I added a test creating a queue filled by random elements with the new method and checking its validity. Also added a potential usage (just took randomly first usage of `add`) and it didn't cause any failures. 

Tried to run benchmarks using https://github.com/mikemccand/luceneutil but 
got an error 
```
src/python/competition.py", line 302
    raise ValueError(f'mode must be "cpu" or "heap" but got: {mode}')
```
may I ask is there any detailed guide on running Lucene benchmarks? 

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.

BaurzhanSakhariev committed:
Implement method to add all collection elements to a PriorityQueue

dweiss commented:
Maybe add a method called addAll(Collection<T>) which would be smart enough to detect when it can use this accelerated addition instead of element-by-element addition? I'm not sure an additional constructor is needed here.

BaurzhanSakhariev commented:
> Maybe add a method called addAll(Collection) which would be smart enough to detect when it can use this accelerated addition instead of element-by-element addition? I'm not sure an additional constructor is needed here.

Good point with `Collection`, thanks. No need to convert list to array if we don't use it directly anyway (because of 1-index shift) - can rather copy from collection.

It's always possible to build a heap if elements are provided and comparable - in JDK they check what kind of collection is provided and take comparator from it - but here we have `lessThan`. My understanding is that user decides whether it's possible to use faster addition - one of use cases when `add()` are not interrupted by `pop()` or any other structure changing operations, i.e when we build first and only then do something. 

I added a constructor as other option would be to use existing ones and then do smth like `addAll(items)` - but I wanted to avoid usage of default constructor, filling it with sentinel null objects. But I actually found an issue in my constructor - need to add `maxSize >= ArrayUtil.MAX_ARRAY_LENGTH` check there, thanks! 

Would it it be fine if I keep the constructor but adjust it to copy directly from `Collection` and add a size check?

dweiss commented:
> I wanted to avoid usage of default constructor, filling it with sentinel null objects

In all honesty, this is probably not something you can ever, ever notice when you compare the cost of building the heap with clearing a chunk of memory. And it makes the API simpler in my opinion. But sure, you can keep the constructor.

BaurzhanSakhariev commented:
Fair point, I changed it to `addAll`. Also added a case to add into partially filled queue - just realized that you probably meant this use case and I misunderstood it.

dweiss commented on lucene/core/src/test/org/apache/lucene/util/TestPriorityQueue.java:
```diff
@@ -163,6 +163,43 @@ public void testInsertWithOverflow() {
+      pq.add(random().nextInt());
+    }
+    pq.addAll(list);
+    pq.checkValidity();
```
Wouldn't it be good to cross-check against some other implementation of a PQ (or even a set) and make sure the elements held in both are the same? I'd also pull out random() outside any loop - this call is costly (the returned Random is cheap).

BaurzhanSakhariev replied:
I added comparison with JDK `PriorityQueue` as implementations are same and internal states are comparable. 

Wanted to compare with Lucene` PriorityQueue` built by calling `add` in loop:  their internal states were different but both were valid as one can build many valid heaps from the same array (for example, 1,2,3, 1 is root and children: 2/3 or 3/2) - didn't add case with `checkValidity` as it's already tested in other tests.

Not sure how compare with `Set `- internally they are different, the only similarity is that removing top element `size` times gets sorted array - but in case of Set it's already tested API, in case of Lucene PQ - validity is tested and other stuff is also tested given that current PQ implementation works fine for heaps with valid internal array.

dweiss replied:
Ah, sorry for not being clearer before - all I meant is to check whether the elements added are actually stored in the PQ (nothing was lost) and that they're in the right order when the PQ is drained. This is effectively equal to what you wrote but instead of comparing arrays, you'd need to drain the PQ (element ordering) and then compare against a (sorted) list of reference elements that you expect the PQ to hold. These two should be equal. I agree that comparing heap element-by-element may be too strict as the implementations may be different (but correct).

BaurzhanSakhariev replied:
done. I didn't remove internal array comparison with jdkQueue - if it's comparable probably good to have extra reassurance?

dweiss replied:
I'd remove it. JDKs can change over time and (in theory) can be implemented by different vendors - I wouldn't tie the test to a particular reference implementation.

BaurzhanSakhariev replied:
removed. I was going to say that `heapify` must be same regardless of the vendor as it's defined algorithm - but then realized that JDK already differs (no index shift, 0-based) 😄

dweiss commented on lucene/core/src/test/org/apache/lucene/util/TestPriorityQueue.java:
```diff
@@ -163,6 +164,57 @@ public void testInsertWithOverflow() {
+    pq.checkValidity();
+    assertOrderedWhenDrained(pq, list);
+  }
+
```
I think you'll have to run 'gradlew tidy' to clean up formatting here.

BaurzhanSakhariev replied:
> Looks good to me, thank you! Let's wait for checks to pass. Please add CHANGES.txt entry under the corresponding version (main or 9x, if you'd like this backported).

Thanks a lot for reviewing!  

Added a CHANGES entry (without Jira issue as there is no one) to Lucene 9.2.0 under Improvements sections (since there is only one usage and there are more add-in-loop usages - so another PR with changing those could be Optimization).

I wanted to run a benchmark (saw nice tables with baseline comparison in other PR-s but couldn't do the same, 
found out https://github.com/mikemccand/luceneutil but had some issues with running it. Probably not necessary anyway as I haven't replaced all other potential usages. 

I don't think I can create issues in JIRA, but if there will be any to apply this method where possible I can address it (at least partially as it's probably big change).

Thanks again for the quick feedback!

BaurzhanSakhariev replied:
Hm, tests passed locally, now some failing - probably there is some flakiness, need to check it

dweiss replied:
> fails on Windows machine because I have CRLF in all files

Don't do this, it's a nightmare. git checkout as-is and commit as-is (this can be configured in git for windows).

BaurzhanSakhariev replied:
looks like there are more things to make it work on windows (gradlew.bat check complains to the lack of perl) - I would rather hold on while I get another machine working - probably it's just local issue but would be better to run all checks.

dweiss commented on lucene/core/src/test/org/apache/lucene/util/TestPriorityQueue.java:
```diff
@@ -163,6 +164,57 @@ public void testInsertWithOverflow() {
+    assertOrderedWhenDrained(pq, oneByOne);
+  }
+
+  public void testAddAllDontFitIntoQueue() {
```
```suggestion
  public void testAddAllDoesNotFitIntoQueue() {
```

dweiss reviewed (commented):
Looks good to me, thank you! Let's wait for checks to pass. Please add CHANGES.txt entry under the corresponding version (main or 9x, if you'd like this backported).

dweiss committed:
Tidy up the changes message (jira issue), change the exception thrown from IllegalArgumentException to ArrayIndexOutOfBoundsException for consistency with add().

dweiss commented:
I've added a proper issue number (https://issues.apache.org/jira/browse/LUCENE-10494) and tweaked minor details of the implementation (exception types thrown in addAll and add were different). I think it's good to go.

dweiss commented:
Thank you!

BaurzhanSakhariev commented:
Thanks!
